### PR TITLE
[MASTER] remove decorator api.returns

### PR DESCRIPTION
**Pull request by Elkasitu**
Related to https://github.com/odoo/enterprise/pull/5081

### CODE REVIEW

File: addons/website_sale/static/tests/tours/website_sale_shop_cart_recovery.js [hunk 1]
```diff
@@ -31,8 +31,7 @@ tour.register('shop_cart_recovery', {
                 'method': 'sale_get_order',
                 'args': [[$('html').data('website-id')]],
             }).then(function (data) {
-                var orderId = parseInt(data.match(/sale\.order\((\d+),\)/)[1]);
```
Comment (mart-e) [inline addons/website_sale/static/tests/tours/website_sale_shop_cart_recovery.js]: wow

Comment (Elkasitu) [reply addons/website_sale/static/tests/tours/website_sale_shop_cart_recovery.js]: best hack odoo 2019

**Comment by xmo-odoo** [comment]
There's a bit of an issue in that it completely breaks BC with methods previously returning a single id, which now return lists instead.

**Comment by mart-e** [comment]
It seems that main `api.returns` are on `copy`. Maybe do the same as in `_call_kw_model_create`.

https://github.com/odoo/odoo/blob/d5d2f5014a9bcf7014901b7099839e1f63ae4d57/odoo/api.py#L350

`message_post` is probably also worth a backward compatibility but not sure how,,,

**Comment by Elkasitu** [comment]
Let's wait for @rco-odoo's feedback as I don't see a way of making this change backwards-compatible without breaking consistency, from my pov it's either-or

File: addons/mail/models/mail_channel.py [hunk 2]
```diff
@@ -351,7 +351,6 @@ def _extract_moderation_values(self, message_type, **kwargs):
                 moderation_status = 'pending_moderation'
         return moderation_status, email
 
-    @api.returns('mail.message', lambda value: value.id)
     def message_post(self, message_type='notification', **kwargs):
```
Comment (rco-odoo) [inline addons/mail/models/mail_channel.py]: The calls to `message_post` from the JS must be adapted, including the mocks of `message_post` in JS tests.

File: odoo/models.py [hunk 3]
```diff
@@ -4314,7 +4309,6 @@ def get_trans(field, old, new):
                     vals_list.append(vals)
                 Translation._upsert_translations(vals_list)
 
-    @api.returns('self', lambda value: value.id)
     def copy(self, default=None):
```
Comment (rco-odoo) [inline odoo/models.py]: Have you checked the server mocks in JS tests?

**Comment by rco-odoo** [comment]
@Elkasitu see my comments above

**Review by MiquelRForgeFlow — COMMENTED** [review]
I think you should delete the `'_returns'` in this line https://github.com/odoo/odoo/blob/master/odoo/api.py#L68. A bit of cleanup of comments there also.

**Comment by Elkasitu** [comment]
@mreficent good catch, thanks

**Comment by mreficent** [comment]
Conlicts

File: odoo/tools/serialization.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    dispatch = dict(xmlrpc.client.Marshaller.dispatch)
+
+    def _Marshaller__dump(self, value, write):
+        # we cannot create a hook for recordsets like for datetime and date objects, because
+        # a recordset of say 'res.users' is not of the same type as a recordset of 'sale.order'
+        # and the __dump function checks for type() and not for isinstance() to determine if an
+        # object is marshallable.
+        # to circumvent this, we check if the object is an instance of BaseModel, which applies
+        # to all recordsets, and then we convert it to id
+        if type(value) not in OdooMarshaller.dispatch and isinstance(value, odoo.models.BaseModel):
```
Comment (xmo-odoo) [inline odoo/tools/serialization.py:28]: `OdooMarshaller.dispatch` could just be `self.dispatch`, though at this point the serialization of dates and datetimes might as well be moved here without bothering with `dispatch`.

**Pull request closed without merging**